src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,275 @@
+                  events <- modifyMVar proto $ \p -> stepProtocol p blk
+                  forM_ events (Telemetry.emit telemetry)
+              , dispatchBlockAsync = atomically . writeTBQueue incomingBlocksQueue
+              }
+
+        worker <- Async.async $ do
+                      block  <- atomically $ readTBQueue incomingBlocksQueue
+                      events <- modifyMVar proto $ \p -> stepProtocol p block
+                      forM_ events (Telemetry.emit telemetry)
+        pure (hdl, worker)
+
+    dispose :: (Handle c tx s IO, Async ()) -> IO ()
+    dispose (_, worker) =
+        -- TODO(adn) Log/emit that we are stopping the protocol runner.
+        Async.cancel worker
+
+-- | Main powerhouse of the 'Protocol' module. In order, this endless
+-- loop does the following:
+--
+-- 1. Gets incoming blocks from the outside;
+-- 2. Performs full (extrinsic) block validation;
+-- 3. Stores validated blocks into the BlockStore or as orphans;
+-- 4. Performs chain selection.
+-- /NOTE/: This function doesn't perform basic validation, and it assumes
+-- the incoming block has been already \"pre-validated\" at the gossip level.
+stepProtocol
+    :: forall c tx s m.
+       ( Ord (BlockHash c)
+       , Ord s
+       , HasHashing c
+       , Buildable (Hash c)
+       , Monad m
+       )
+    => Protocol c tx s m
+    -> Block c tx (Sealed c s)
+    -> m (Protocol c tx s m, [NotableEvent])
+stepProtocol mgr incomingBlock = do
+    -- Step 2. Try to store the block; if this is recognised as to be an
+    -- orphan, we store it immediately. If this block is scheduled for
+    -- inclusion to extend the tip of the chain, then its full validity is
+    -- enforced.
+    extendsTip <- do
+        currentTip <- BlockStore.getTip bsPublicAPI
+        pure $ blockHash currentTip == Block.parentHash incomingBlock
+
+    (mgr', evts) <-
+       if extendsTip
+          then do
+              -- Get the last (cached) 'mutableChainDepth' blocks and performs
+              -- full block validation.
+              let depth = fromIntegral $ mutableChainDepth (protoConfig mgr)
+              ancestors <- BlockStore.getBlocksByDepth bsPublicAPI depth
+              case protoValidateFull mgr (toNewestFirst ancestors) incomingBlock of
+                  Left validationError ->
+                      let evt = BlockValidationFailedEvent (blockHash incomingBlock) validationError
+                      in pure (mgr, [evt])
+                  Right () -> do
+                      -- Step 3: store the fully-validated block.
+                      BlockStore.insertBlock bsPrivateAPI incomingBlock
+                      pure (mgr, mempty)
+
+          -- NOTE(adn) This is a side effect of the fact that storing
+          -- the genesis as an orphan would insert it into the orphanage with
+          -- its parent hash, which means that we won't be able to retrieve
+          -- a potentiol candidate branching off from genesis in our 'selectBestChain'
```

Comment:
```suggestion
          -- a potential candidate branching off from genesis in our 'selectBestChain'
```

Reply:
I think it's time for me to start coding with `:set spell` always on 😂

Reply:
a42f62e

Reply:
I was going to suggest that :joy: